sui-execution/latest/sui-adapter/src/package_unification.rs

```diff
@@ -0,0 +1,379 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use move_binary_format::{binary_config::BinaryConfig, file_format::Visibility};
+use sui_types::{
+    base_types::{ObjectID, SequenceNumber},
+    move_package::{normalize_module, MovePackage},
+    storage::BackingPackageStore,
+    transaction::{Command, ProgrammableTransaction},
+    type_input::TypeInput,
+    Identifier,
+};
+
+pub struct PTBLinkageMetadata {
+    pub types: Vec<ObjectID>,
+    pub entry_functions: Vec<ObjectID>,
+    pub non_entry_functions: Vec<ObjectID>,
+    pub publication_linkages: Vec<ObjectID>,
+    pub upgrading_packages: Vec<ObjectID>,
+    pub all_packages: BTreeMap<ObjectID, MovePackage>,
+}
+
+pub struct LinkageConfig {
+    pub fix_top_level_functions: bool,
+    pub fix_types: bool,
+}
+
+#[derive(Debug)]
+pub enum ConflictResolution {
+    Exact(SequenceNumber, ObjectID),
+    AtLeast(SequenceNumber, ObjectID),
+    Never(ObjectID),
+}
+
+impl LinkageConfig {
+    pub fn strict() -> Self {
+        Self {
+            fix_top_level_functions: true,
+            fix_types: true,
+        }
+    }
+
+    pub fn loose() -> Self {
+        Self {
+            fix_top_level_functions: false,
+            fix_types: false,
+        }
+    }
+}
+
+impl ConflictResolution {
+    pub fn unify(&mut self, other: &ConflictResolution) -> anyhow::Result<()> {
```

Comment:
Nit, I'd prefer 
```suggestion
    pub fn unify(&self, other: &ConflictResolution) -> anyhow::Result<ConflictResolution> {
```

Reply:
I always find these sort of mutable unifications harder to follow (but maybe thats just me)

Reply:
FWIW I significantly prefer them due to their efficiency improvements. In this instance, switching to return the value would require us to update the BTreeMap _every_ time we do unification, whereas now we just update the RHS.